app/serialization.py

```diff
@@ -89,50 +89,81 @@ def deserialize_host_xjoin(data):
-    if "reporter" in fields:
-        serialized_host["reporter"] = host.reporter
-    if "stale_timestamp" in fields:
-        serialized_host["stale_timestamp"] = stale_timestamp and _serialize_datetime(stale_timestamp)
-    if "stale_warning_timestamp" in fields:
-        serialized_host["stale_warning_timestamp"] = stale_timestamp and _serialize_datetime(stale_warning_timestamp)
-    if "culled_timestamp" in fields:
-        serialized_host["culled_timestamp"] = stale_timestamp and _serialize_datetime(culled_timestamp)
-        # without astimezone(timezone.utc) the isoformat() method does not include timezone offset even though iso-8601
-        # requires it
-    if "created" in fields:
-        serialized_host["created"] = _serialize_datetime(host.created_on)
-    if "updated" in fields:
-        serialized_host["updated"] = _serialize_datetime(host.modified_on)
-    if "tags" in fields:
-        serialized_host["tags"] = _serialize_tags(host.tags)
-    if "system_profile" in fields:
-        serialized_host["system_profile"] = host.system_profile_facts or {}
+def serialize_host(host, staleness_timestamps, fields=DEFAULT_FIELDS, canonical_fields=_CANONICAL_FACTS_FIELDS):
+    return {
+        **serialize_canonical_facts(host.canonical_facts, canonical_fields),
+        **_serialize_host_fields(host, fields, staleness_timestamps),
+    }
+
+
+def serialize_host_sparse(host, staleness_timestamps, sparse_fieldset):
+    host_fields = {"id"}
+    canonical_fields = set()
+
+    if "host" in sparse_fieldset:
+        requested_host_fields = set(sparse_fieldset["host"].split(","))
+        canonical_fields |= requested_host_fields & _CANONICAL_FACTS_FIELDS
+        host_fields |= requested_host_fields - canonical_fields - {"system_profile"}
+
+    serialized_host = serialize_host(host, staleness_timestamps, host_fields, canonical_fields)
+
+    if "system_profile" in sparse_fieldset:
+        system_profile_attributes = sparse_fieldset["system_profile"].split(",")
+        serialized_host["system_profile"] = {}
+        for system_profile_attribute in system_profile_attributes:
+            if system_profile_attribute not in host.system_profile_facts:
+                continue
+            if host.system_profile_facts and system_profile_attribute in host.system_profile_facts:
+                serialized_host["system_profile"][system_profile_attribute] = host.system_profile_facts[
+                    system_profile_attribute
+                ]
+            else:
+                serialized_host["system_profile"][system_profile_attribute] = None
 
     return serialized_host
 
 
+def _serialize_host_fields(host, fields, staleness_timestamps):
+    def raw(field):
+        return getattr(host, field)
+
+    def function(field, serialize):
+        if field == "created":
+            field = "created_on"
+        elif field == "updated":
+            field = "modified_on"
+
+        value = getattr(host, field)
```

Comment:
If we want to get rid of the ifs and of the argument name override, we can use mapping like this:

```suggestion
        field_map = {"created": "created_on", "updated": "modified_on"}
        attr = field_map.get(field, field)
        value = getattr(host, field)
```